nmdc_server/crud.py

```diff
@@ -488,3 +489,90 @@ def release_submission_lock(db: Session, submission_id: str):
         raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Submission not found")
     submission.locked_by = None  # type: ignore
     db.commit()
+
+
+#############################
+# SUBMISSION ACCESS CONTROL #
+#############################
+read_roles = [
+    models.SubmissionEditorRole.editor,
+    models.SubmissionEditorRole.metadata_contributor,
+    models.SubmissionEditorRole.owner,
+    models.SubmissionEditorRole.viewer,
+]
+
+metadata_edit_roles = [
+    models.SubmissionEditorRole.editor,
+    models.SubmissionEditorRole.metadata_contributor,
+    models.SubmissionEditorRole.owner,
+]
+
+context_edit_roles = [
+    models.SubmissionEditorRole.editor,
+    models.SubmissionEditorRole.owner,
+]
+
+contributors_edit_roles = [
+    models.SubmissionEditorRole.owner,
+]
+
+
+def can_read(db: Session, submission_id: str, user_orcid: str) -> Optional[bool]:
+    role = (
+        db.query(models.SubmissionRole)
+        .filter(
+            and_(
+                models.SubmissionRole.user_orcid == user_orcid,
+                models.SubmissionRole.submission_id == submission_id,
+            )
+        )
+        .first()
+    )
+    submission: Optional[models.SubmissionMetadata] = db.query(models.SubmissionMetadata).get(
+        submission_id
+    )
+    if submission is None:
+        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Submission not found")
+    is_creator = submission.author_orcid == user_orcid
+    return (is_creator or (role and models.SubmissionEditorRole(role.role) in read_roles)) is True
+
+
+def can_edit_all(db: Session, submission_id: str, user_orcid: str) -> Optional[bool]:
```

Comment:
I'd either change the function name to be longer (probably preferred if possible) or add a comment describing the function intent. "can edit all" I think means they can edit the entire submission, maybe "can_edit_entire_submission"?

Reply:
I updated this function name and `can_read` above in commit 92950c9